clang/test/SemaSYCL/deferred-diagnostics-emit.cpp

```diff
@@ -0,0 +1,85 @@
+// RUN: %clang_cc1 -I %S/Inputs -triple spir64 -fsycl-is-device -verify -fsyntax-only %s
+
+//
+//    ensuring that the SYCL diagnostics that are typically deferred, correctly emit 
+//
+
+#include <sycl.hpp>
+
+struct S {
+  virtual void foo() {}
+};
+
+int calledFromKernel(int a){
+  // expected-error@+1 {{zero-length arrays are not permitted in C++}}
+  int MalArray[0];  
+
+  // expected-error@+1 {{__float128 is not supported on this target}}
+  __float128 malFloat = 40;  
+
+  // not sure if 'no virtual function' is a _deferred_ diagnostic, testing anyway 
```

Comment:
What exactly do you mean here?
BTW, we diagnose only usages of virtual functions, as SYCL spec requires. Using of classes with virtual functions is not prohibited.

Reply:
I am not sure if this particular violation ( a virtual function in SYCL context ) results in a _deferred_ diagnostic, as opposed to an _immediate_ diagnostic.   @erichkeane  added it to one of the files when he was working on reducing the problem. So I kept it in the test.

Reply:
AFAIK we diagnose virtual functions usage through SemaSYCL callgraph traversing. See https://github.com/intel/llvm/blob/f87291d3b61a5d4b54e4265948868c7deada9dfc/clang/lib/Sema/SemaSYCL.cpp#L233 . But, I suppose, we will move this to Sema and deferred diagnostics.

But we already diagnose variadic function calls through deferred diagnostic, I remember patch from Erich doing this https://github.com/intel/llvm/pull/998 .
So If here you are trying to check only deferred diagnostics, I suggest switch from virtual functions to variadic functions.

Reply:
Ok, I switched to variadic.